sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -287,76 +289,145 @@ private SpecializedBlobClientBuilder prepareBuilder() {
-                    parallelTransferOptions.getNumBuffers(), parallelTransferOptions.getProgressReceiver());
-
-            // See ProgressReporter for an explanation on why this lock is necessary and why we use AtomicLong.
-            AtomicLong totalProgress = new AtomicLong(0);
-            Lock progressLock = new ReentrantLock();
-
-            // Validation done in the constructor.
-            UploadBufferPool pool = new UploadBufferPool(finalParallelTransferOptions.getNumBuffers(),
-                finalParallelTransferOptions.getBlockSize());
-
-            /*
-            Break the source Flux into chunks that are <= chunk size. This makes filling the pooled buffers much easier
-            as we can guarantee we only need at most two buffers for any call to write (two in the case of one pool
-            buffer filling up with more data to write). We use flatMapSequential because we need to guarantee we
-            preserve the ordering of the buffers, but we don't really care if one is split before another.
-             */
-            Flux<ByteBuffer> chunkedSource = data
-                .filter(ByteBuffer::hasRemaining)
-                .flatMapSequential(buffer -> {
-                    if (buffer.remaining() <= finalParallelTransferOptions.getBlockSize()) {
-                        return Flux.just(buffer);
-                    }
-                    int numSplits =
-                        (int) Math.ceil(buffer.remaining() / (double) finalParallelTransferOptions.getBlockSize());
-                    return Flux.range(0, numSplits)
-                        .map(i -> {
-                            ByteBuffer duplicate = buffer.duplicate().asReadOnlyBuffer();
-                            duplicate.position(i * finalParallelTransferOptions.getBlockSize());
-                            duplicate.limit(Math.min(duplicate.limit(),
-                                (i + 1) * finalParallelTransferOptions.getBlockSize()));
-                            return duplicate;
-                        });
-                });
+                parallelTransferOptions.getNumBuffers(), parallelTransferOptions.getProgressReceiver());
+
+            BlockBlobAsyncClient blockBlobAsyncClient = getBlockBlobAsyncClient();
+            return determineUploadFullOrChunked(data, (stream) -> uploadInChunks(blockBlobAsyncClient, stream,
+                finalParallelTransferOptions, headers, metadata, tier, accessConditionsFinal),
+                (stream, length) -> blockBlobAsyncClient.uploadWithResponse(
+                    ProgressReporter.addProgressReporting(stream, finalParallelTransferOptions.getProgressReceiver()),
+                    length, headers, metadata, tier, null, accessConditionsFinal));
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    private Mono<Response<BlockBlobItem>> uploadInChunks(BlockBlobAsyncClient blockBlobAsyncClient,
```

Comment:
It looks as if there were many perf optimizations made in this method. Are these not applicable to `uploadBlocks()`, the similar method used when we upload from a file and so know the length already? Or are all the optimizations based on the scenario of not having this information present? If they're truly incompatible then so be it.

Reply:
The methods are fairly similar in name and parameters aside from that they differ largely in their intended use. `uploadBlocks` (maybe rename it to something mentioning files) will read a file from disk and upload it using a cold (replayable) flux, `uploadInChunks` in meant to be used by hot (non-replayable) fluxes which need to have an intermediate buffering applied. Having `uploadFromFile` call into `uploadInChunks` would lead to additional buffering overhead that isn't needed. I think there are definitely pieces that could be take out of both methods and put somewhere common but it's out of scope for this change, luckily this would be an implementation change and could be done in the future without breaking API.

Reply:
Perhaps some added inline docs to state the intent and effect of some of the less intuitive steps would help clarify this.